File: spec.maven.plugin/src/main/java/org/spek/maven/plugin/SpekMojo.java
package org.spek.maven.plugin;

import org.apache.maven.artifact.DependencyResolutionRequiredException;
import org.apache.maven.plugin.AbstractMojo;
import org.apache.maven.plugins.annotations.LifecyclePhase;
import org.apache.maven.plugins.annotations.Mojo;
import org.apache.maven.plugins.annotations.Parameter;
import org.apache.maven.plugins.annotations.ResolutionScope;
import org.apache.maven.project.MavenProject;
import org.spek.console.cmd.CmdPackage;
import org.spek.console.cmd.Options;

import java.io.File;
import java.net.MalformedURLException;
import java.net.URL;
import java.net.URLClassLoader;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

@Mojo(name = "spek", requiresDependencyResolution = ResolutionScope.TEST, defaultPhase = LifecyclePhase.TEST)
public class SpekMojo extends AbstractMojo {

    @Parameter(defaultValue = "", property = "packageName")
    private String packageName = "";

    @Parameter(defaultValue = "true", property = "text")
    private Boolean text = true;

    @Parameter(defaultValue = "false", property = "html")
    private Boolean html = false;

    @Parameter(property = "outputFile")
    private String outputFile = null;

    @Parameter(property = "cssFile")
    private String cssFile = null;

    @Parameter(defaultValue = "${project}", required = true, readonly = true)
    private MavenProject project;

    @Override
    public void execute() {

        updateContextClassLoader();

        Options options = new Options(
                packageName,
                text,
                html,
                (outputFile != null ? outputFile : ""),
                (cssFile != null ? cssFile : ""));

        CmdPackage.runSpecs(options);
    }

    private void updateContextClassLoader() {
        try {
            Set<URL> urls = new HashSet<>();
            List<String> elements = project.getTestClasspathElements();
            for (String element : elements) {
                urls.add(new File(element).toURI().toURL());
            }

            ClassLoader contextClassLoader = URLClassLoader.newInstance(
                    urls.toArray(new URL[0]),
                    Thread.currentThread().getContextClassLoader());

            Thread.currentThread().setContextClassLoader(contextClassLoader);

        } catch (DependencyResolutionRequiredException e) {
            throw new RuntimeException(e);
        } catch (MalformedURLException e) {
            throw new RuntimeException(e);
        }
    }
}

File: spec.maven.plugin/src/main/kotlin/org/spek/maven/plugin/SpekMojo.kt
package org.spek.maven.plugin

import org.apache.maven.plugin.AbstractMojo
import org.apache.maven.project.MavenProject
import org.spek.console.cmd.Options
import java.net.URL
import java.io.File
import java.net.URLClassLoader
import org.apache.maven.plugins.annotations.ResolutionScope
import org.apache.maven.plugins.annotations.LifecyclePhase
import org.apache.maven.plugins.annotations.Mojo as mojo
import org.apache.maven.plugins.annotations.Parameter as parameter
import org.spek.console.cmd.runSpecs

mojo(name = "spek", requiresDependencyResolution = ResolutionScope.TEST, defaultPhase = LifecyclePhase.TEST)
public open class SpekMojo: AbstractMojo() {
    parameter(defaultValue = "", property = "packageName")
    private var packageName: String = ""

    parameter(defaultValue = "true", property = "text")
    private var text: Boolean = true

    parameter(defaultValue = "false", property = "html")
    private var html: Boolean = false

    parameter(property = "outputFile")
    private var outputFile: String? = null

    parameter(property = "cssFile")
    private var cssFile: String? = null

    parameter(defaultValue = "\${project}", required = true, readonly = true)
    private var project: MavenProject? = null

    public override fun execute(): Unit {
        runWithProjectClassPath {
            runSpecs(Options(packageName, text, html, outputFile ?: "", cssFile ?: ""))
        }
    }

    private fun runWithProjectClassPath(action: () -> Unit) {
        val classLoader = Thread.currentThread().getContextClassLoader()!!

        var elements = project?.getTestClasspathElements()!!

        val urls = Array<URL>(elements.size, { File(elements.get(it).toString()).toURI().toURL() })

        val newClassLoader = URLClassLoader.newInstance(urls, classLoader)!!

        Thread.currentThread().setContextClassLoader(newClassLoader)
        try {
            action()
        } finally {
            Thread.currentThread().setContextClassLoader(classLoader)
        }
    }
}
